sdk/eventhub/Azure.Messaging.EventHubs/tests/Consumer/EventHubConsumerClientLiveTests.cs

```diff
@@ -2060,200 +1589,479 @@ await using (var consumer = new EventHubConsumerClient(EventHubConsumerClient.De
+            var result = new ReadState();
 
-            await foreach (var partitionEvent in readAwaitable)
+            try
             {
-                if (iterationCallback != null)
+                var readAwaitable = consumer.ReadEventsFromPartitionAsync(partition, startingPosition.Value, readOptions, cancellationToken).ConfigureAwait(false);
+
+                await foreach (var partitionEvent in readAwaitable)
                 {
-                    if (await iterationCallback().ConfigureAwait(false))
+                    // Track the events as they are read.
+
+                    if (partitionEvent.Data == null)
                     {
-                        break;
+                        ++result.EmptyCount;
                     }
-                }
-
-                if (partitionEvent.Data != null)
-                {
-                    readEvents.Add(partitionEvent);
-                    consecutiveEmpties = 0;
-                }
-                else if (++consecutiveEmpties >= consecutiveEmptyDelayThreshold)
-                {
-                    if (consecutiveEmpties >= consecutiveEmptyLimit)
+                    else
                     {
-                        break;
+                        var eventId = partitionEvent.Data.Properties[EventGenerator.IdPropertyName].ToString();
+
+                        if ((result.Events.TryAdd(eventId, partitionEvent)) && (result.Events.Count >= expectedEventCount))
+                        {
+                            break;
+                        }
                     }
 
-                    if (readEvents.Count < expectedEventCount)
+                    // If there's a callback registered per-tick, invoke it and respect its
+                    // decision on whether iteration should continue.
+
+                    if ((iterationCallback != null) && (!(await iterationCallback(result).ConfigureAwait(false))))
                     {
-                        await Task.Delay(consecutiveEmptyDelayMilliseconds, cancellationToken).ConfigureAwait(false);
+                        break;
                     }
                 }
+
+                // Delay for a moment before returning the results to ensure that cleanup has registered
+                // with the service and the associated link is no longer alive.
+
+                await Task.Delay(250).ConfigureAwait(false);
+            }
+            catch (TaskCanceledException)
+            {
+                // The test should assert on the cancellation token; treat this as
+                // expected and don't bubble.
             }
 
-            return readEvents;
+            return result;
         }
 
         /// <summary>
-        ///   Reads from all partitions until there are no events available for
-        ///   a number of consecutive iterations.
+        ///   Reads the events across all partitions from a given position, tracking the operation.
         /// </summary>
-        ///
```

Comment:
I believe this line removal was an accident.

Reply:
Indeed it was.